### ISSUE TITLE
Too many chunks due to uncounted header

### ISSUE THREAD

**Issue by OverkillGuy** [OverkillGuy/qrxfil#3]
A file that can be split into 10 chunks of 1KB of data will actually need more, because the strict 1KB QR payload includes the header, currently 6 bytes, say 02OF16. This is fine, except when splitting larger files where the 6 bytes add up to > 1KB (also, passing the 100 = 3 digits makes this into 8 bytes).

This means the following happens:

```bash
$ git bundle create jibyconf_all.bundle -b master
$ qrxfil exfil jibyconf_all.bundle jibyconf_all_exfil
File jibyconf_all.bundle. base64 size: 671628 bytes = 656 chunks of 1KB
Saved QR 01/656
Saved QR 02/656
Saved QR 03/656
Saved QR 04/656
Saved QR 05/656
...
Saved QR 653/656
Saved QR 654/656
Saved QR 655/656
Saved QR 656/656
Saved QR 657/656
Saved QR 658/656
Saved QR 659/656
Saved QR 660/656
Split file in 660 QR chunks, in folder "jibyconf_all_exfil"
```

And taking those chunk shows prefix is `660OF656`. The extra chunks are to carry all the headers.

Let's adapt the chunk_total estimation to take into account header prefixes.

**Comment by OverkillGuy** [comment]
> what is the value of having "chunk x of y" instead of just "chunk x". Does knowing the total number in advance help you reassemble them in anyway?

Valid question! 
I think I care yes, because I may have 76 chunks, but need to know I'm DONE with no 77th chunk lost. This means any chunk we recover tells us how many to expect.

> If not, then you never need to estimate the total number before hand (or have the entire file in memory at once), you just iterate through and create chunks until you run out of file. That would simplify the problem, no?

I'm already managing to avoid having file entirely in memory by using buffered reads with 1KB increments, converting it to base64 into another file. The size measurement is done by using cursor position at the end of base64 version to count filesize, which I reach naturally after this conversion.
Also, the QR payload is the order of KB and I don't expect anyone to scan more than 100 QR codes, much less 1000, we can reasonably bound filesize in the megabytes range, nothing crazy memory-wise (though I do intend to keep it functional and performant if possible.

> FYI, the problem of estimating the total number of items in a set, based on the numbers you've seen _so far_ is called "The German Tank Problem"

Good catch, though I'm lucky I don't need crazy heuristics, because base64 encoding has a flat encoding rate of 4/3rds of original filesize + between 1-3 chars of flat padding, so I can use that as upper bound, but even then, I _know_ filesize because I measure it.

**Comment by danieleades** [comment]
I see! I was wondering why you buffered to another file first...

**Comment by danieleades** [comment]
I thought of another way to solve this. What if you add a footer to the *last* chunk? That way you know when you're missing chunks, *and* you can process everything in a single pass without buffering.
The only thing you don't know for sure is your progress when you're part way through decoding chunks, but that's where the heuristic comes in.

**Comment by OverkillGuy** [comment]
Related, I am currently considering making filename + file sha256 (integrity) a part of the payload, which means the last chunk will have useful info about restoration without overburdening the header with redundant bytes. This would fit in that kind of payload info, but brings me to "real file format territory" (parsing rules etc).

I hadn't considered making number of chunks a last-page thing, but am quite attached to the progress bar of sorts. Note that the intention is for fully out of order scanning to be supported, like reverse order.

I expect (for the first drafts) at most 999 chunks, so header is `999OF999` = 8 bytes off of 1024bytes, not too crazy (though I am considering too a kind of file format's magic number, almost like a MIMETYPE. Will see).

The easiest way to keep the chunk total count consistent would be to insert chunk_count at the last minute after having COUNTED them, rather than computed. Will see if this is feasible. First I need to refactor the code to be more unit testable.

**Comment by danieleades** [comment]
Would be easy enough to make the behaviour configurable, too

**Comment by OverkillGuy** [comment]
Quick note on working around this: I just scanned 581 chunks (which pretended to be 576 initially, due to this bug).

The workaround for this was to just update the chunk total in the scanned text file with the real number of chunks (upmost file number in QR PNG exfil folder):

```bash
sed -i -E 's/([0-9]{3})OF576/\1OF581/g' file_to_restore.txt
```

**Comment by danieleades** [comment]
Would be interesting to know what the performance implications are.

With the current implementation you making 2 passes through the data, and buffering to a file, so you're filesystem IO-bound.

Got a funny feeling you'll more than double the speed of you can do it in one pass

**Comment by OverkillGuy** [comment]
Right. So. 

I agree on chunking iterator (though I was curious/nervous about it at first, it makes more and more sense in the code the more I look at it, decouple the file open from the streaming, better for tests), though of course it's a nice to have, not requirement.

The root cause of the double-write (first file -> base64 file, then base64 -> chunk files) is because I needed to know how many chunks would there be in total (which as discussed above is crucial for the usecase). But! `base64` encoding is deterministic in encoded size, equal to `(4 / 3) * original + padding`, with `padding` of 0-3 bytes (represented as `=`) depending on `size %4`.

So, to avoid double writes, I can calculate exact base64 size in new `fn base64_size_estimator(original_size: u64) -> u64`, as well as a higher level `fn chunk_count_calculator(original_size: u64) -> u64` which incidentally can be instrumented to predict the overhead of chunks, which is the issue this bug is about (and have nice unit tests of it), and get a single pass through the file to exfiltrate, which is wrapped in a base64EncoderWriter anyway so it's all pretty.

So by getting a chunk iterator, and doing some pre-calculation, I can get cleaner code, more performance, and more accuracy!
I'll get on this bug then, because it's literally the only thing preventing me from tagging a `v0.1.0` (see [milestone](https://github.com/OverkillGuy/qrxfil/milestone/1)).

... Also note that performance-wise, it is MUCH slower to generate a QR code + save image file than any straightforward <1Mb base64 encoding file->file, so we're not gaining a lot there ;)

**Comment by OverkillGuy** [comment]
Shelving this for a bit while I fix the bug #17. Will be back soon enough!
